dcap/types/src/quote3.rs

```diff
@@ -44,34 +77,217 @@ impl<T: AsRef<[u8]>> Quote3<T> {
+    // This is the `size` field, *not* the length of `bytes`.
+    data_size: usize,
+}
 
-    #[cfg(feature = "alloc")]
-    use std::vec;
+impl<'a> TryFrom<&'a [u8]> for AuthenticationData<'a> {
+    type Error = Error;
+    fn try_from(bytes: &'a [u8]) -> Result<Self> {
+        let bytes_length = bytes.len();
+
+        let data_size = u16_from_bytes(bytes, 0)? as usize;
+
+        let needed_size = data_size + mem::size_of::<u16>();
+        if bytes_length < needed_size {
+            Err(Error::InvalidInputLength {
+                required_size: needed_size,
+                actual_size: bytes_length,
+            })
+        } else {
+            Ok(Self { bytes, data_size })
+        }
+    }
+}
+
+impl<'a> AuthenticationData<'a> {
+    pub fn size(&self) -> usize {
+        self.data_size + mem::size_of::<u16>()
+    }
+}
+
+/// The Quoting enclave certification data
+///
+/// Table 9 of
+/// <https://download.01.org/intel-sgx/latest/dcap-latest/linux/docs/Intel_SGX_ECDSA_QuoteLibReference_DCAP_API.pdf>.
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+struct CertificationData<'a> {
+    bytes: &'a [u8],
+    // Since these are small and the `size` has to be read we always unpack
+    // for availability.
+    data_type: u16,
+    // This is the size of the data field, *not* the size of `bytes`
+    data_size: usize,
+}
+
+impl<'a> TryFrom<&'a [u8]> for CertificationData<'a> {
+    type Error = Error;
+    fn try_from(bytes: &'a [u8]) -> Result<Self> {
+        let bytes_length = bytes.len();
+
+        // type (2 bytes) + size (4 bytes)
+        let mut needed_size = mem::size_of::<u16>() + mem::size_of::<u32>();
+
+        if bytes_length < needed_size {
+            return Err(Error::InvalidInputLength {
+                required_size: needed_size,
+                actual_size: bytes_length,
+            });
+        }
+
+        // These shouldn't fail since we ensured the length up above
+        let data_type = u16_from_bytes(bytes, 0)?;
+        let data_size = u32_from_bytes(bytes, mem::size_of::<u16>())? as usize;
+
+        needed_size += data_size;
+        if bytes_length < needed_size {
+            Err(Error::InvalidInputLength {
+                required_size: needed_size,
+                actual_size: bytes_length,
+            })
+        } else {
+            Ok(Self {
+                bytes,
+                data_type,
+                data_size,
+            })
+        }
+    }
+}
+
+fn u32_from_bytes(bytes: &[u8], offset: usize) -> Result<u32> {
+    const SIZE: usize = mem::size_of::<u32>();
+    let mut u32_bytes = [0u8; SIZE];
+    let end = offset + SIZE;
+    let copy_bytes = bytes.get(offset..end).ok_or(Error::InvalidInputLength {
+        required_size: end,
+        actual_size: bytes.len(),
+    })?;
+    u32_bytes.copy_from_slice(copy_bytes);
+    Ok(u32::from_le_bytes(u32_bytes))
+}
 
+fn u16_from_bytes(bytes: &[u8], offset: usize) -> Result<u16> {
```

Comment:
same deal as above, It could be my lack of rust knowledge but I couldn't get a generic implementation working due to `from_le_bytes()` wanting to know what the type is before hande